### PR TITLE
simplest implementation of promisc flag on open_link, works on Linux

### PR DESCRIPTION
**Pull request by gabrik**
Simple implementation of the promisc parameter on open_link
to enable promiscuos mode on a interface while opening it, tested on 
Ubuntu 16.04.5
Ocaml 4.07.1
dune 1.6.3

See https://github.com/haesbaert/rawlink/issues/12

Signed-off-by: gabrik <gabriele.baldoni@gmail.com>

### CODE REVIEW

File: lib/rawlink_stubs.c [hunk 1]
```diff
@@ -271,40 +264,58 @@ af_packet_setfilter(int fd, value vfilter)
...
+int af_packet_set_promisc(int fd, int promisc, char *ifname)
+{
+	struct ifreq ifopts;
+	int r;
+	if (!promisc)
+		return (1);
+
+	strncpy(ifopts.ifr_name, ifname, IFNAMSIZ - 1);
+	caml_enter_blocking_section();
+	r = ioctl(fd, SIOCGIFFLAGS, &ifopts);
```
Comment (haesbaert) [inline lib/rawlink_stubs.c]: You have to test the error of each syscall and abort, also ifopts should be bzroed before.

File: lib/rawlink_stubs.c [hunk 2]
```diff
@@ -271,40 +264,58 @@ af_packet_setfilter(int fd, value vfilter)
 	return (r);
 }
 
+int af_packet_set_promisc(int fd, int promisc, char *ifname)
+{
+	struct ifreq ifopts;
+	int r;
```
Comment (haesbaert) [inline lib/rawlink_stubs.c]: new line after declarations, also keep same indentation style.

**Review by haesbaert — COMMENTED** [review]
I'll pick this up and give you credit, will try to push until tomorrow, need to test on bpf first.
many thanks !

File: lib/rawlink.ml [hunk 3]
```diff
@@ -35,16 +35,16 @@ type driver =
...
 external dhcp_client_filter: unit -> string = "caml_dhcp_client_filter"
 external driver: unit -> driver = "caml_driver"
 external unix_bytes_read: Unix.file_descr -> Cstruct.buffer -> int -> int -> int =
   "lwt_unix_bytes_read"
 external bpf_align: int -> int -> int = "caml_bpf_align"
 
-let open_link ?filter ifname =
-  { fd = opensock ?filter:filter ifname;
+let open_link ?promisc ?filter ifname =
+  { fd = opensock ?promisc:promisc ?filter:filter ifname;
```
Comment (hannesm) [inline lib/rawlink.ml]: FWIW, you can drop `:promisc` and `:filter` here:
```OCaml
let fd = opensock ?promisc ?filter ifname in
{ fd ; packets = ref [] ; buffer = Cstruct.create 65536 }
```

**Comment by gabrik** [comment]
Hi both, I have updated the PR addressing the comments, the indentation style should be the right one, so feel free to borrow this last version for the feature branch, also I have a question on the Lwt version, why Lwt_rawlink.open_link return t instead of returing t Lwt.t ?

**Comment by haesbaert** [comment]
I merged something inspired on this, thanks for the report.
See b7284a9aecf3da5799ac258a4407df7a3fc16447

**Pull request closed without merging**